### DISCUSSION TITLE
How to prepare OpenGL context (also working Qt python example)

### DISCUSSION THREAD

**Discussion by JeanElsner** [google-deepmind/mujoco#214] [Asking for Help]
This probably isn't a MuJoCo issue. We're trying to render MuJoCo scenes within a Qt GUI using the python bindings provided by PySide6. There is an OpenGL [Widget](https://doc.qt.io/qt-5/qopenglwidget.html) that already prepares an OpenGL context etc. Essentially this works fine, there is however an issue with shadows. Once shadows are activated other features like the skybox, haze and transparent geoms disappear from the rendered output. It seems that the render passes are still computed (runtime increases) but not displayed on the OpenGL surface. I tried to play around with some OpenGL settings to no avail. As far as I can tell the custom `mjrContext` is the same as for glfw, which has no issues.

There might be an easy fix for this, if you're an OpenGL expert and/or understand how to prepare the context for MuJoCo :-)
In any case I provide a minimal working example (except for the shadow issue) which some of you may find useful.

https://gist.github.com/JeanElsner/755d0feb49864ecadab4ef00fd49a22b
![qt_reflection](https://user-images.githubusercontent.com/19858553/160579163-df1635ca-21ef-4c2d-8bc5-d1bfe1d16b93.png)
![qt_shadow](https://user-images.githubusercontent.com/19858553/160579198-9ffa66c2-a6a2-4d29-8c04-cbb254a2db67.png)

**Comment by JeanElsner** [comment]
Not sure what's going on here. However, if you use [QOpenGLWindow](https://doc.qt.io/qt-6/qopenglwindow.html) or just a QWindow and add the OpenGL surface manually, everything works out. I updated the example.

![pyside6](https://user-images.githubusercontent.com/19858553/161110030-fdf14645-10f5-44f4-9f87-2418a0565067.png)

**Comment by saran-t** [comment]
Hi, sorry I haven't been replying, but that's mostly because I don't really have anything to suggest. That's still the case, but I'd encourage you to see if the OpenGL changes that we made in 2.1.4 made any difference for you. (I'd be surprised if that's the case, but worth trying anyway I think.)

**Comment by hesic73** [comment]
Thank you for sharing! I met with the same issue. `QOpenGLWindow` works, but `QOpenGLWidget` doesn't work.
